### DISCUSSION TITLE
Table : some cells are not detected

### DISCUSSION THREAD

**Discussion by BaptCha** [jsvine/pdfplumber#808] [Ask for help with specific PDFs]
Hello,

First, thank you for this awesome library !

I have an issue with a table (where some cells are not detected properly ?) as you can see bellow:

<img width="627" alt="Screenshot 2023-02-09 at 16 43 23" src="https://user-images.githubusercontent.com/34271863/217862412-f915222b-5593-448d-ad66-6174a80f91f4.png">

The PDF is [here](https://www.ema.europa.eu/en/documents/product-information/eliquis-epar-product-information_en.pdf) page 13.

I tried all table settings without success :/

Would you mind offering some ideas on how to solve this problem? Thanks so much for your kind help!

**Comment by samkit-jain** [comment]
Hi @BaptCha Appreciate your interest in the library. For the PDF you shared, you can use the table settings as
```python
{
    "vertical_strategy": "explicit",
    "horizontal_strategy": "explicit",
    "explicit_vertical_lines": p.curves+p.edges,
    "explicit_horizontal_lines": p.curves + p.edges,
}
```
where `p` is the page object. `p = pdf.pages[12]`.

Following is the output I get
![output](https://user-images.githubusercontent.com/15127115/218015564-0a9cf7d1-1ddd-4170-8199-f33f4a4de1a0.png)

```
['System organ class', 'Prevention of \nVTE in adult \npatients who \nhave undergone \nelective hip or \nknee replacement \nsurgery (VTEp)', 'Prevention of \nstroke and \nsystemic \nembolism in \nadult patients \nwith NVAF, with \none or more risk \nfactors (NVAF)', 'Treatment of DVT \nand PE, and \nprevention of \nrecurrent DVT and \nPE (VTEt)']
['Blood and lymphatic system disorders', None, None, None]
['Anaemia', 'Common', 'Common', 'Common']
['Thrombocytopenia', 'Uncommon', 'Uncommon', 'Common']
['Immune system disorders', None, None, None]
['Hypersensitivity, allergic oedema and \nAnaphylaxis', 'Rare', 'Uncommon', 'Uncommon']
['Pruritus', 'Uncommon', 'Uncommon', 'Uncommon*']
['Angioedema', 'Not known', 'Not known', 'Not known']
['Nervous system disorders', None, None, None]
['Brain haemorrhage†', 'Not known', 'Uncommon', 'Rare']
['Eye disorders', None, None, None]
['Eye haemorrhage (including \nconjunctival haemorrhage)', 'Rare', 'Common', 'Uncommon']
['Vascular disorders', None, None, None]
['Haemorrhage, haematoma', 'Common', 'Common', 'Common']
['Hypotension (including procedural \nhypotension)', 'Uncommon', 'Common', 'Uncommon']
['Intra-abdominal haemorrhage', 'Not known', 'Uncommon', 'Not known']
['Respiratory, thoracic and mediastinal disorders', None, None, None]
['Epistaxis', 'Uncommon', 'Common', 'Common']
['Haemoptysis', 'Rare', 'Uncommon', 'Uncommon']
['Respiratory tract haemorrhage', 'Not known', 'Rare', 'Rare']
['Gastrointestinal disorders', None, None, None]
['Nausea', 'Common', 'Common', 'Common']
['Gastrointestinal haemorrhage', 'Uncommon', 'Common', 'Common']
['Haemorrhoidal haemorrhage', 'Not known', 'Uncommon', 'Uncommon']
['Mouth haemorrhage', 'Not known', 'Uncommon', 'Common']
['Haematochezia', 'Uncommon', 'Uncommon', 'Uncommon']
['Rectal haemorrhage, gingival \nbleeding', 'Rare', 'Common', 'Common']
['Retroperitoneal haemorrhage', 'Not known', 'Rare', 'Not known']
['Hepatobiliary disorders', None, None, None]
['Liver function test abnormal, asparate \naminotransferase increased, blood \nalkaline phosphatase increased, blood \nbilirubin increased', 'Uncommon', 'Uncommon', 'Uncommon']
['Gamma-glutamyltransferase \nincreased', 'Uncommon', 'Common', 'Common']
['Alanine aminotransferase increased', 'Uncommon', 'Uncommon', 'Common']
['Skin and subcutaneous tissue disorders', None, None, None]
```

**Reply by BaptCha** [reply]
Hi @samkit-jain,

That solved it !
Thanks you very much, you made my day.

Thanks again for this awesome lib !

**Comment by BaptCha** [comment]
Hi again :) 

The preceding solution is working great but I get some wrong results as you can see bellow:
```python
{
    "vertical_strategy": "explicit",
    "horizontal_strategy": "explicit",
    "explicit_vertical_lines": p.curves+p.edges,
    "explicit_horizontal_lines": p.curves + p.edges,
}
```
<img width="645" alt="Screenshot 2023-02-16 at 16 03 15" src="https://user-images.githubusercontent.com/34271863/219401955-6240eb85-ed8b-4687-90de-9e44861e4074.png">

I do not care about the figure at the top but I am trying to get the table at the bottom.
Somehow the figure edges and curves at the top are screwing up the table parsing at the bottom.


Using default parameters or keeping the proposed one with cropping the page around the table solve it.
That being said I am looking to find a general solution that works everywhere without human intervention.


If anyone has any ideas in mind, I would be very grateful.


The PDF is [here](https://www.ema.europa.eu/en/documents/product-information/levemir-epar-product-information_fr.pdf#page=11) page 11.

**Reply by samkit-jain** [reply]
For this page, using just the default 
```python
{
    "vertical_strategy": "lines",
    "horizontal_strategy": "lines"
}
```
works
![output](https://user-images.githubusercontent.com/15127115/219572386-479c9a3f-337c-4b86-bd99-a3ae72444dfd.png)

```
['Durée de\nl’étude', 'Levemir une fois\npar jour', 'Levemir deux fois\npar jour', 'Insuline NPH', 'Insuline glargine']
['20 semaines', '+0,7 kg', '', '+1,6 kg', '']
['26 semaines', '', '+1,2 kg', '+2,8 kg', '']
['52 semaines', '+2,3 kg', '+3,7 kg', '', '+4,0 kg']
```

Note that there is no one-size-fits-all table settings so one configuration might work for a few tables but may fail for others.

**Reply by BaptCha** [reply]
Thank again @samkit-jain !

Unfortunately I was looking for a general solution that works everywhere :sweat_smile: 
Well I'll try to find the best parameters for the most documents.

**Reply by zangar111** [reply]
.

![image](https://github.com/user-attachments/assets/0417f939-2252-444d-b5b8-8e6c0b737ecf)
![image](https://github.com/user-attachments/assets/d253ca48-b586-4fae-aef1-948b4e81cb7f)


Hi, I have a problem. Why does it jump from 002781123 directly to 002965650? I tried using your settings, but it didn’t work.

**Reply by zangar111** [reply]
@BaptCha @samkit-jain

**Reply by jsvine** [reply]
Hi @zangar111, can you provide the PDF and a short Python script that reproduces problem you're seeing?

**Reply by zangar111** [reply]
> Hi @zangar111, can you provide the PDF and a short Python script that reproduces problem you're seeing?

Thank you for the quick response

pdf: https://t.me/uto92/2519

Here is the code: https://t.me/+LWwlj5_8BLM1MjQy

**Reply by zangar111** [reply]
@jsvine

**Reply by jsvine** [reply]
Hi @zangar111, can you provide the code/PDF directly here, rather than via Telegram links? Alternatively, you can email them to me at the address listed in my profile.

**Reply by zangar111** [reply]
> Hi @zangar111, can you provide the code/PDF directly here, rather than via Telegram links? Alternatively, you can email them to me at the address listed in my profile.

Do you have another way to communicate?

**Comment by kalle07** [comment]
hey next problem ... i tryed a lot with strategy and snap but nothing work for both tables in one document (a book)
maybe you have any idea

![grafik](https://github.com/user-attachments/assets/b2937202-c568-4b45-a9e3-95582aef9a00)
this one works with standart (also with strategy vertical-lines, horizontal-lines)

![grafik](https://github.com/user-attachments/assets/cba478c7-e14f-4a77-9c87-40d60c465528)
this one works none standart (only with strategy vertical-text, horizontal-lines), but than the other table are broken.

the problem of the second one are the missing left and right outlines, if i draw them all is fine, with outline also standard work, but i mean should be easy recognized without i think  ;)

so i need an option where are booth fine . . . thx for help

[table_my01.pdf](https://github.com/user-attachments/files/20027817/table_my01.pdf)

**Reply by jsvine** [reply]
`pdfplumber`'s general philosophy leans toward explicitness and less toward mimicking how humans "see" tables. So in this case, the `pdfplumber` solution to the second example provided here would be to programmatically determine where you "see" those right and left borders, and to pass them to the `explicit_vertical_lines` and `explicit_horizontal_lines` table extraction settings.

**Reply by kalle07** [reply]
thx, someone is a live  ;)
i thought lines were recognized by plumber, also lines and edges you can snap and join ...
so maybe easy to close all open lines, rectangels by plumber !?! 
i have no idea to dynapical use explicit_vertical_lines.

**Comment by kalle07** [comment]
any hints ?
